app/queues/common/expiration-request.server.ts

```diff
@@ -0,0 +1,98 @@
+import { Worker, Queue, UnrecoverableError } from 'bullmq';
+
+import { redis } from '~/lib/redis.server';
+import logger from '~/lib/logger.server';
+import { prisma } from '~/db.server';
+
+import { addNotification } from '../notifications/notifications.server';
+import { deleteDnsRequest } from '../dns/delete-record-flow.server';
+import { RecordType } from '../notifications/expiration-notification.server';
+
+import type { ExpirationStatusPayload as ExpirationRequestPayload } from '../notifications/expiration-notification.server';
+
+const { EXPIRATION_REPEAT_FREQUENCY, JOB_REMOVAL_FREQUENCY } = process.env;
+const JOB_REMOVAL_INTERVAL = 7 * 24 * 60 * 60 * 1000;
+declare global {
+  var __expiration_request_init__: boolean;
+}
+// queue name
+const expirationRequestQueueName = 'expiration-request';
+
+// query to fetch expired records from database based on type
+const getExpiredRecords = (type: RecordType) => {
+  switch (type) {
+    case RecordType.DnsRecord:
+      return prisma.record.findMany({
+        where: {
+          expiresAt: {
+            lt: new Date(),
+          },
+        },
+        include: {
+          user: true,
+        },
+      });
+    case RecordType.Certificate:
+      logger.info(`TODO: return expired certificates from the db`);
+      break;
+    default:
+      throw new Error(`Invalid recordType: ${type}`);
+  }
+};
+
+// queue initialization
+const expirationRequestQueue = new Queue<ExpirationRequestPayload>(expirationRequestQueueName, {
+  connection: redis,
+});
+
+export function addExpirationRequest(type: RecordType) {
+  const jobName = `expirationRequestQueueName-${type}`;
+  return expirationRequestQueue.add(
+    jobName,
+    { type },
+    {
+      repeat: { every: Number(EXPIRATION_REPEAT_FREQUENCY) || 24 * 60 * 60 * 1000 },
+      removeOnComplete: { age: Number(JOB_REMOVAL_FREQUENCY) || JOB_REMOVAL_INTERVAL },
+      removeOnFail: { age: Number(JOB_REMOVAL_FREQUENCY) || JOB_REMOVAL_INTERVAL },
+    }
+  );
+}
+
+// worker definition
+const expirationRequestWorker = new Worker<ExpirationRequestPayload>(
+  expirationRequestQueueName,
+  async (job) => {
+    const { type } = job.data;
+    switch (type) {
+      case RecordType.DnsRecord:
+        try {
+          logger.info('process DNS record expiration');
+          let dnsRecords = await getExpiredRecords(type);
```

Comment:
So this can return a non-array?  That makes sense why you were wanting to check the `dnsRecords?.length`:

```ts
const dnsRecords = await getExpiredRecords(type);
if(!dnsRecords) {
  return;
}
return Promise.all(dnsRecords.map(...))
```

Reply:
Yes, it can be potentially null or undefined.